wasi/wasi.go

```diff
@@ -396,13 +424,48 @@ func proc_exit(*wasm.HostFunctionCallContext, uint32) {
 	// not implemented yet
 }
 
-func environ_sizes_get(*wasm.HostFunctionCallContext, uint32, uint32) (err Errno) {
-	// not implemented yet
-	return
+// environ_sizes_get is a WASI API that returns the number of the environment variables and the total buffer size that
+// environ_get API will require to store the value of the environment variables.
+// * environCountPtr: a pointer to an address of uint32 type. The number of the environment variable is written there.
+// * environBufSizePtr: a pointer to an address of uint32 type. The total size of the buffer that the environment variable data requires is written there.
+//
+// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-environ_sizes_get---errno-size-size
+func (w *WASIEnvironment) environ_sizes_get(ctx *wasm.HostFunctionCallContext, environCountPtr uint32, environBufSizePtr uint32) (err Errno) {
+	if !ctx.Memory.PutUint32(environCountPtr, uint32(len(w.environ.nullTerminatedValues))) {
+		return EINVAL
+	}
+	if !ctx.Memory.PutUint32(environBufSizePtr, w.environ.totalBufSize) {
+		return EINVAL
+	}
+
+	return ESUCCESS
 }
 
-func environ_get(*wasm.HostFunctionCallContext, uint32, uint32) (err Errno) {
-	return
+// environ_get is a WASI API to read the environment variable data.
+// * environPtr:
+//     A pointer to a buffer. environ_get writes multiple *C.char pointers in sequence there. In other words, this is an array of *char in C.
+//     The pointers point to environment variables. Each environment variable is a key/value pair joined with `=`, and it is a null-terminated string.
+//     The number of the environment variables matches the value that environ_sizes_get returns.
+// * environBufPtr:
+//     A pointer to a buffer. envrion_get writes the environment variables as null-terminated strings in the given buffer.
+//     The total number of bytes written there is the value that args_sizes_get returns in argsBufSizePtr. The caller must ensure that
+//     the buffer has the enough size.
+//     Each *C.char pointer that can be obtained from environPtr points to the beginning of each of these null-terminated strings.
```

Comment:
do we think `*C.char pointer` will be understandable and relevant in go code, given choices of how to describe this?